pkg/receiver/oracledbreceiver/scraper.go

```diff
@@ -0,0 +1,316 @@
+	sessionExecuteCountSQL      = "select sum(se.value) as VALUE  from v$session ss, v$sesstat se, v$statname sn where se.STATISTIC# = sn.STATISTIC# and NAME = 'execute count'"
+	sessionParseCountTotalSQL   = "select sum(se.value) as VALUE from v$session ss, v$sesstat se, v$statname sn where se.STATISTIC# = sn.STATISTIC# and NAME = 'parse count (total)'"
+	sessionUserCommitsSQL       = "select sum(se.value) as VALUE from v$session ss, v$sesstat se, v$statname sn where se.STATISTIC# = sn.STATISTIC# and NAME = 'user commits'"
+	sessionUserRollbacksSQL     = "select sum(se.value) as VALUE from v$session ss, v$sesstat se, v$statname sn where se.STATISTIC# = sn.STATISTIC# and NAME = 'user rollbacks'"
+	sessionCountSQL             = "select status, type, count(*) as VALUE FROM v$session GROUP BY status, type"
+	systemResourceLimitsSQL     = "select RESOURCE_NAME, CURRENT_UTILIZATION, MAX_UTILIZATION, CASE WHEN TRIM(INITIAL_ALLOCATION) LIKE 'UNLIMITED' THEN '-1' ELSE TRIM(INITIAL_ALLOCATION) END as INITIAL_ALLOCATION, CASE WHEN TRIM(LIMIT_VALUE) LIKE 'UNLIMITED' THEN '-1' ELSE TRIM(LIMIT_VALUE) END as LIMIT_VALUE from v$resource_limit"
+	tablespaceUsageSQL          = "select TABLESPACE_NAME, BYTES from DBA_DATA_FILES"
+	tablespaceMaxSpaceSQL       = "select TABLESPACE_NAME, (BLOCK_SIZE*MAX_EXTENTS) AS VALUE FROM DBA_TABLESPACES"
+)
+
+type dbProviderFunc func() (*sql.DB, error)
+
+type clientProviderFunc func(*sql.DB, string, *zap.Logger) dbClient
+
+type scraper struct {
+	logger             *zap.Logger
+	metricsBuilder     *metadata.MetricsBuilder
+	dbProviderFunc     dbProviderFunc
+	clientProviderFunc clientProviderFunc
+	db                 *sql.DB
+	id                 config.ComponentID
+	instanceName       string
+	scrapeCfg          scraperhelper.ScraperControllerSettings
+	startTime          pcommon.Timestamp
+	metricsSettings    metadata.MetricsSettings
+}
+
+func newScraper(id config.ComponentID, metricsBuilder *metadata.MetricsBuilder, metricsSettings metadata.MetricsSettings, scrapeCfg scraperhelper.ScraperControllerSettings, logger *zap.Logger, providerFunc dbProviderFunc, clientProviderFunc clientProviderFunc, instanceName string) *scraper {
+	return &scraper{
+		id:                 id,
+		metricsBuilder:     metricsBuilder,
+		metricsSettings:    metricsSettings,
+		scrapeCfg:          scrapeCfg,
+		logger:             logger,
+		dbProviderFunc:     providerFunc,
+		clientProviderFunc: clientProviderFunc,
+		instanceName:       instanceName,
+	}
+}
+
+var _ scraperhelper.Scraper = (*scraper)(nil)
+
+func (s *scraper) ID() config.ComponentID {
+	return s.id
+}
+
+func (s *scraper) Start(context.Context, component.Host) error {
+	s.startTime = pcommon.NewTimestampFromTime(time.Now())
+	var err error
+	s.db, err = s.dbProviderFunc()
+	if err != nil {
+		return fmt.Errorf("failed to open db connection: %w", err)
+	}
+	return nil
+}
+
+func (s *scraper) Scrape(ctx context.Context) (pmetric.Metrics, error) {
```

Comment:
What do you think about the number of lines of code in this method? Would it be better to break it up and see if boilerplate can be factored out? (Can be done in a separate PR.) Also, more generally, I think a lot of this logic could be done once, at Start() time, rather than on each scrape. The sqlquery receiver, for example, configures and registers the `dbClient`s* and scrapers at start time, one for each query. In any case, doesn't have to be done now IMO since I think this provides functionality that customers are looking for.

*terrible name (my mistake); should be called statement runners or something

Reply:
No worries - changed to initialize in Start, good call.